pinot-transport/src/main/java/com/linkedin/pinot/routing/builder/LargeClusterRoutingTableBuilder.java

```diff
@@ -0,0 +1,299 @@
+    // The algorithm to generate a routing table is thus:
+    // 1. Compute the inverse external view, a mapping of servers to segments
+    // 2. For each routing table to generate:
+    //   a) Pick MAX_SERVERS_PER_QUERY distinct servers
+    //   b) Check if the server set covers all the segments; if not, add additional servers until it does.
+    //   c) Order the segments in our server set in ascending order of number of replicas present in our server set
+    //   d) For each segment, pick a random replica with proper weighting
+    //   e) Return that routing table
+    //
+    // Given that we can generate routing tables at will, we then generate many routing tables and use them to optimize
+    // according to two criteria: the variance in workload per server for any individual table as well as the variance
+    // in workload per server across all the routing tables. To do so, we generate an initial set of routing tables
+    // according to a per-routing table metric and discard the worst routing tables.
+
+    RoutingTableGenerator routingTableGenerator = new RoutingTableGenerator();
+    routingTableGenerator.init(externalView, instanceConfigList);
+
+    PriorityQueue<Pair<Map<String, Set<String>>, Float>> topRoutingTables = new PriorityQueue<>(ROUTING_TABLE_COUNT, new Comparator<Pair<Map<String, Set<String>>, Float>>() {
+      @Override
+      public int compare(Pair<Map<String, Set<String>>, Float> left, Pair<Map<String, Set<String>>, Float> right) {
+        // Float.compare sorts in ascending order and we want a max heap, so we need to return the negative of the comparison
+        return -Float.compare(left.getValue(), right.getValue());
+      }
+    });
+
+    for (int i = 0; i < ROUTING_TABLE_COUNT; i++) {
+      topRoutingTables.add(generateRoutingTableWithMetric(routingTableGenerator));
+    }
+
+    // Generate routing more tables and keep the ROUTING_TABLE_COUNT top ones
+    for(int i = 0; i < (ROUTING_TABLE_GENERATION_COUNT - ROUTING_TABLE_COUNT); ++i) {
+      Pair<Map<String, Set<String>>, Float> newRoutingTable = generateRoutingTableWithMetric(routingTableGenerator);
+      Pair<Map<String, Set<String>>, Float> worstRoutingTable = topRoutingTables.peek();
+
+      // If the new routing table is better than the worst one, keep it
+      if (newRoutingTable.getRight() < worstRoutingTable.getRight()) {
+        topRoutingTables.poll();
+        topRoutingTables.add(newRoutingTable);
+      }
+    }
+
+    // Return the best routing tables
+    List<ServerToSegmentSetMap> routingTables = new ArrayList<>(topRoutingTables.size());
+    while(!topRoutingTables.isEmpty()) {
+      Pair<Map<String, Set<String>>, Float> routingTableWithMetric = topRoutingTables.poll();
+      routingTables.add(new ServerToSegmentSetMap(routingTableWithMetric.getKey()));
+    }
+
+    return routingTables;
+  }
+
+  private Pair<Map<String, Set<String>>, Float> generateRoutingTableWithMetric(RoutingTableGenerator routingTableGenerator) {
```

Comment:
suggest having a separate method to compute metric given a routing table. We can then test the metric method independently.